expression/constant.go

```diff
@@ -245,7 +348,21 @@ func (c *Constant) EvalTime(ctx sessionctx.Context, _ chunk.Row) (val types.Time
 
 // EvalDuration returns Duration representation of Constant.
 func (c *Constant) EvalDuration(ctx sessionctx.Context, _ chunk.Row) (val types.Duration, isNull bool, err error) {
-	if c.DeferredExpr != nil {
+	if c.DeferredParam != nil {
+		dt := c.DeferredParam.Value()
+		if dt.IsNull() {
+			return types.Duration{}, true, err
+		}
+		val, err := dt.ToString()
+		if err != nil {
+			return types.Duration{}, true, err
+		}
+		dur, err := types.ParseDuration(ctx.GetSessionVars().StmtCtx, val, types.MaxFsp)
```

Comment:
It's better to wrap a `cast string as duration` function on this deferred param. A method is to transform the deferred param to a `expression.Constant`, construct a unfolded scalar `Cast` function upon that `Constant`.

Reply:
good catch, can we do this on next PR to avoid useless `str<->duration`?

we also need change https://github.com/pingcap/tidb/blob/cb23b524accce9ff1d1168acb0b5a102f3e7a22c/server/conn_stmt.go#L538 to let `dt` be a duration datum, also some other types